### PR TITLE
Bvh kokkos view

### PR DESCRIPTION
**Pull request by Rombur**
For now keep a struct TreeConstruction and don't put TreeConstruction in details. This will need to be done later.

### CODE REVIEW

File: packages/Kokkos/src/DTK_LinearBVH_def.hpp [hunk 1]
```diff
@@ -1,92 +1,176 @@
...
+    KOKKOS_INLINE_FUNCTION
+    void operator()( int const i ) const { _indices[i] = _max; }
+
+  private:
+    unsigned int const _max;
+    Kokkos::View<unsigned int *, DeviceType> _indices;
+};
+
+template <typename NO>
+class SetIndices
```
Comment (dalg24) [inline packages/Kokkos/src/DTK_LinearBVH_def.hpp:37]: This is the equivalent of `std::iota`.

File: packages/Kokkos/src/DTK_LinearBVH_def.hpp [hunk 2]
```diff
@@ -1,92 +1,176 @@
...
+template <typename NO>
+class SetBoundingBoxes
+{
+  public:
+    using DeviceType = typename NO::device_type;
+    using ExecutionSpace = typename DeviceType::execution_space;
+
+    SetBoundingBoxes( Kokkos::View<Node *, DeviceType> leaf_nodes,
+                      Kokkos::View<int *, DeviceType> indices,
+                      AABB const *bounding_boxes )
```
Comment (dalg24) [inline packages/Kokkos/src/DTK_LinearBVH_def.hpp]: `Kokkos::View`?

File: packages/Kokkos/src/DTK_LinearBVH_def.hpp [hunk 3]
```diff
@@ -1,92 +1,176 @@
...
+    Kokkos::fence();
+    tree_construction.assignMortonCodes( bounding_boxes, morton_indices, n,
+                                         _internal_nodes[0].bounding_box );
+
     // sort them along the Z-order space-filling curve
-    for ( int i = 0; i < n; ++i ) // parallel_for
-        _indices[i] = i;
-    Details::sortObjects( morton_indices.data(), _indices.data(), n );
+    Functor::SetIndices<NO> set_indices_functor( _indices );
+    Kokkos::parallel_for( "set_indices",
```
Comment (dalg24) [inline packages/Kokkos/src/DTK_LinearBVH_def.hpp:110]: `iota`

File: packages/Kokkos/src/details/DTK_DetailsAlgorithms.hpp [hunk 4]
```diff
@@ -22,6 +23,74 @@ void expand( Box &box, Box const &other );
...
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void join( volatile Box &dst, volatile Box const &src ) const
+    {
+        for ( int d = 0; d < 3; ++d )
+        {
+            // We need to access the underlying element directly because
+            // operator[] is not volatile
+            dst._minmax.m_elem[2 * d] = KokkosHelpers::min(
```
Comment (dalg24) [inline packages/Kokkos/src/details/DTK_DetailsAlgorithms.hpp]: That's problematic.  It was renamed to `m_internal_implementation_private_member_data ` in the Kokkos develop branch.

File: packages/Kokkos/test/tstDetailsTreeConstruction.cpp [hunk 5]
```diff
@@ -40,45 +40,56 @@ TEUCHOS_UNIT_TEST( DetailsBVH, morton_codes )
...
     // solution
-    std::vector<unsigned int> k = {2, 1, 4, 3};
+    //
+    using DeviceType = typename NO::device_type;
+    int constexpr n = 4;
+    Kokkos::View<unsigned int *, DeviceType> k( "k", n );
+    k[0] = 2;
+    k[1] = 1;
+    k[2] = 4;
+    k[3] = 3;
```
Comment (dalg24) [inline packages/Kokkos/test/tstDetailsTreeConstruction.cpp]: Is it possible to allocate a vector with the initializer_list and wrap a view around it?

File: packages/Kokkos/test/tstDetailsTreeConstruction.cpp [hunk 6]
```diff
@@ -106,46 +117,61 @@ TEUCHOS_UNIT_TEST( DetailsBVH, number_of_leading_zero_bits )
...
+    fi[6] = 8;
+    fi[7] = 13;
+    fi[8] = 21;
+    fi[9] = 34;
+    fi[10] = 55;
+    fi[11] = 89;
+    fi[12] = 144;
+
+    DataTransferKit::TreeConstruction<NO> tc;
+    TEST_EQUALITY( tc.commonPrefix( fi, n, 0, 0 ), 32 + 32 );
```
Comment (dalg24) [inline packages/Kokkos/test/tstDetailsTreeConstruction.cpp]: `commonPrefix` is static so we don't actually need an object do we?

Comment (Rombur) [reply packages/Kokkos/test/tstDetailsTreeConstruction.cpp]: No we don't need one. I first created the structure and afterwards I made all the functions static. So we can choose if we want to use the structure or not.

File: packages/Kokkos/test/tstDetailsTreeConstruction.cpp (hunk 5 above)
Comment (dalg24) [reply packages/Kokkos/test/tstDetailsTreeConstruction.cpp]:
```
std::vector<unsigned int> k = {2, 1, 4, 3};
Kokkos::View<unsigned int *, DeviceType, Kokkos::MemoryUnmanaged>(..., k.data(), k.size());
```

**Pull request merged**